src/main/scala/abalone/variant/Variant.scala

```diff
@@ -102,6 +103,144 @@ abstract class Variant private[variant] (
         )
     }.toMap
 
+  def validMovesOf2(situation: Situation, validMoves1: Map[Pos, List[Move]]): Map[Pos, List[(String, Move)]] =
+    situation.board.piecesOf(situation.player).flatMap {
+      case ((pos, piece)) => {
+        Map(
+          pos -> pos.neighbours.flatMap {
+            case Some(neighbour) if(situation.board.piecesOf(situation.player).contains(neighbour)) => {
+              neighbour.neighbours.flatMap {
+                case Some(neighbourOfNeighbour)
+                  if (situation.board.isEmptySquare(Some(neighbourOfNeighbour))
+                  && (pos.isInLine(neighbour, neighbourOfNeighbour) || situation.board.isEmptySquare(pos.dir(neighbour.dir(neighbourOfNeighbour))))
+                  && !validMoves1.get(pos).toList.flatMap(_.map(_.toUci)).exists(m => m.toString == s"Move(${pos},${neighbourOfNeighbour},None)")) =>
+                    Some( ("nopush", Move(piece, pos, neighbourOfNeighbour, situation, boardAfter(situation, neighbour, neighbourOfNeighbour), false)) )
+                case Some(neighbourOfNeighbour)
+                  if (pos.isInLine(neighbour, neighbourOfNeighbour) && situation.board.piecesOf(!situation.player).contains(neighbourOfNeighbour)) =>
+                    if (neighbourOfNeighbour.dir(pos.dir(neighbour)) == None)
+                      Some( ("pushout", Move(piece, pos, neighbourOfNeighbour, situation, boardAfter(situation, neighbour, neighbourOfNeighbour), false, neighbour.dir(pos.dir(neighbour)))) )
+                    else
+                      if (situation.board.isEmptySquare(neighbourOfNeighbour.dir(pos.dir(neighbour))))
+                        Some( ("push", Move(piece, pos, neighbourOfNeighbour, situation, boardAfter(situation, neighbour, neighbourOfNeighbour), false)) )
+                      else None
+                case _ => None
+              }
+            }
+            case _ => None
+          }
+        )
+      }
+    }.toMap
+
+  // def validMovesOf3(situation: Situation, @nowarn validMoves2: Map[Pos, List[(String, Move)]]): Map[Pos, List[(String, Move)]] =
+  //   situation.board.piecesOf(situation.player).flatMap {
```

Comment:
as above, if you are reusing the same code we should refactor into val, e.g. 
```
{
val turnPieces = situation.board.piecesOf(situation.player)
code to return...
}
```

Reply:
do not review movesOf3 it is not ready, should have just not committed sorry

Reply:
but so you mean I should put `val turnPieces = situation.board.piecesOf(situation.player)` into e.g. validMoves() so it can be reused by validMovesOf1 and validMovesOf2 at least for example ?